kep/61-real-load-aware-scheduling/README.md

```diff
@@ -0,0 +1,608 @@
+2. The default in-tree scheduling plugins (Score) that Kubernetes provides don't consider live node utilization values.
+
+This proposal utilizes real time resource usage to schedule pods with proposed plugins. The eventual goal is to increase cluster utilization and decrease costs of cluster management without breaking the contract of Kubernetes Resource Model.
+
+### Goals
+
+1. Provide configurable scheduling plugins to increase cluster utilization.
+2. Expected CPU Utilisation per node should not go beyond X%.
+3. Instances of an app should spread across failure domains. The failure domain would be host at the lowest level and can extend to racks, zones, regions, etc.
+4. To not affect the behavior of default score plugins unless necessary.
+5. Implement the above features as Score plugins
+
+### Non-Goals
+
+1. Implement the above constraints (2, 3) as Filter plugins.
+2. Descheduling due to unexpected outcome (hot nodes, fragmentation etc.) of past scoring by plugins is not addressed in the initial design.
+3. Memory, Network, and Disk utilization are not considered in the initial design.
+
+## Proposal
+
+
+### User Stories
+
+
+#### Story 1
+
+
+As a company relying on the public cloud, we would like to minimize machine costs by efficiently utilizing the nodes leased.
+
+
+As a company-owned data center, we would like to minimize the hardware and maintenance associated with clusters by getting the maximum possible juice out of existing machines.
+
+
+#### Story 2
+
+Increasing resource utilization as much as possible may not be the right solution for all clusters. As it always takes some time to scale up a cluster to handle the sudden spikes of load, cluster admins would like to leave adequate room for the bursty load, so there is enough time to add more nodes to the cluster.
+
+
+### Notes/Constraints/Caveats
+
+Enabling our plugin(s) will cause conflict with 2 default scoring plugins: "NodeResourcesLeastAllocated" and "NodeResourcesBalancedAllocation" plugins. So it is strongly advised to disable them when enabling plugin(s) mentioned in this proposal.
+For the 3rd Goal, since we utilise PodTopologySpread plugin scoring and to prevent double scoring, it is recommended to disable "PodTopologySpread" Score plugin only. The motivation behind using this plugin is explained under "Spreading with BestFitBinPack".
+
+
+### Risks and Mitigations
+
+If utilization metrics are not available for a long time, we will fall back to the best fit bin pack based on allocations. There is no user action needed for this.
+
+
+## Design Details
+
+Our design consists of the following components as outlined in the diagram and described below. We propose two plugins namely "BestFitBinPack" and "SafeBalancing". Both of them use metrics from load watcher for scoring nodes with different algorithms.
+
+<img src="images/design.png" alt="design" width="720" height="628"/>
```

Comment:
nit:
`![Design](images/design.png)`

Reply:
That doesn't resize the image :), this one is of high quality and default size is quite big.